**Autor**: Gabriel Przytocki <br>
**Data**: 18/07/2021

# Challenge 1: Data Engineer Jr

In [1]:
from google.colab import drive

# Coletando os dados salvos no Drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
# Manipulação de matrizes
import numpy as np

# Load Data

In [3]:
import pandas as pd

authors = 'drive/MyDrive/datasets/mirum-data-engineer-challenge/authors.csv'       #authors.csv
categor = 'drive/MyDrive/datasets/mirum-data-engineer-challenge/categories.csv'    #categories.csv
dataset = 'drive/MyDrive/datasets/mirum-data-engineer-challenge/dataset.csv'       #dataset.csv
formats = 'drive/MyDrive/datasets/mirum-data-engineer-challenge/formats.csv'       #formats.csv

aut = pd.read_csv(authors) #authors.csv
cat = pd.read_csv(categor) #categories.csv
dat = pd.read_csv(dataset) #dataset.csv
fom = pd.read_csv(formats) #formats.csv

In [4]:
# Informações do DataFrame "dataset"
dat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20216 entries, 0 to 20215
Data columns (total 28 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   authors             20216 non-null  object 
 1   bestsellers-rank    9227 non-null   float64
 2   categories          20216 non-null  object 
 3   description         17804 non-null  object 
 4   dimension-x         18937 non-null  float64
 5   dimension-y         18071 non-null  float64
 6   dimension-z         18937 non-null  float64
 7   edition             2157 non-null   object 
 8   edition-statement   5156 non-null   object 
 9   for-ages            201 non-null    object 
 10  format              20023 non-null  float64
 11  id                  20216 non-null  int64  
 12  illustrations-note  6567 non-null   object 
 13  image-checksum      20216 non-null  object 
 14  image-path          20216 non-null  object 
 15  image-url           20216 non-null  object 
 16  impr

# Visualizing data

## `authors.csv`

In [5]:
#authors.csv
aut.head(5)

author_id          author_name
0       3481                  NaN
1        181           - Mimamour
2      11255  - Samoral Helgueras
3       3734        . Macnoughton
4       8644               16.95n

## `categories.csv`

In [6]:
#categories.csv
cat.head(5)

category_id                                category_name
0          176  20th Century & Contemporary Classical Music
1         3291  20th Century & Contemporary Classical Music
2         2659      20th Century History: C 1900  To C 2000
3         2661          21st Century History: From C 2000 -
4         2472                                          ABC

## `formats.csv`

In [7]:
#formats.csv
fom.head(5)

format_id format_name
0         15       Audio
1         18       Board
2          9        Book
3          4          CD
4          2    Calendar

## `dataset.csv`

In [8]:
#dataset.csv
dat.head()

authors  ...  weight
0  [1, 2]  ...  182.00
1     [3]  ...  208.65
2     [4]  ...  361.00
3     [5]  ...  244.00
4     [6]  ...  208.00

[5 rows x 28 columns]

# Preparando os dados

In [9]:
def parse_str_to_list(str_list: str,
                      df_index: str):
    
        """
        Função responsável por transformar uma coluna específica de um DataFrame,
        esta que possui a froma de uma lista, mas é do tipo str (e.g. "[1, 2]"), em
        uma lista real, em que seus itens internos são números inteiros.

        Exemplo:
            input: "[1, 2]"
            output: [1, 2]

        Params:
            str_list: lista no formato str a ser transformada.

            df_index: Nome da coluna no "dataset.csv" para ser consultada no df.

        Returns:
            Retorna uma lista real com seus itens no tipo inteiro dentro.
        """

        # Inicializando a variável final com a lista str transformada em list
        elements = []

        # Iterando sobre os elementos da coluna alvo do "dataset"
        for i in str_list[df_index]:

            # Inicializando a variável acerca do novo elemento
            element = []
            content = i[1:-1] # Excluindo "[" e "]"
            content = content.split(", ") # Separando os itens da lista  por ","

            # Excluindo os elementos nulos e inadequados
            for e in content:
                if e != "" and e != " " and e != None:
                    element.append(int(e)) # Adicionando o elemento como inteiro

            elements.append(element) # Adicionando a nova lista à lista final

        return elements


def parse_index_to_label(
    id_label: str, 
    name_label: str, 
    df_index: str, 
    df: pd.DataFrame, 
    df_legend: pd.DataFrame) -> list:

    """
    Função reponsável por transformar a lista de códigos do "dataset.csv" em uma
    lista com as labels reais correspondentes ao código.

    Exemplo:
        input: "[1, 2]": str
        output: ["Autor A", "Autor B"]: list

    Params:
        id_label: nome da coluna correspondente ao id de um item do "dataset.csv"
                  Contém o mesmo id presente no dataset mencionado.

        name_label: nome da coluna correspondente ao "name" de um item do 
                    "dataset.csv". Contém o mesmo "name" presente no dataset
                    mencionado. 

        df_index: Nome da coluna no "dataset.csv" para ser consultada no df.

        df: DataFrame final a ser consultado para se extrair os códigos, de
            maneira a construir uma nova lista com os "names" em questão.

        df_legend: DataFrame com os names dos ids a serem consultados.

    Returns:
        Retorna uma lista com várias listas dento, cada lista contendo um ou mais
        nomes das labels alvo de um determinado livro (e.g. nomes de autores).
    """
    
    # Utilizando a função 'parse_str_to_list' para obter a lista no formato list
    # Inicializando a variável final com os elementos substituídos.
    elements = parse_str_to_list(df,
                                 df_index)

    # Contador macro
    major_counter = 0
    # Iterando sobre cada lista dentro da lista transformada

    for i in elements:
        # Contador micro
        minor_counter = 0

        # Iterando sobre os elementos de cada lista
        for k in i:
            # Novo elemento da lista, do qual consulta no df com os names e realiza
            # a substituição
            k = df_legend[df_legend[id_label] == k][name_label].values[0]

            # Inclui o novo elemento na variável final
            # Utiliza os contadores para acessar a posição específica
            elements[major_counter][minor_counter] = k

            minor_counter += 1 # Incremento de contador
        major_counter += 1 # Incremento de contador

    return elements

## `authors.csv` + `dataset.csv`: Substituindo index por label

In [10]:
# Lista com os nomes de autores, dos quais subtituem os código de autores
# da coluna "authors" do "dataset"
authors_parsed = parse_index_to_label("author_id",
                                      "author_name",
                                      "authors",
                                      dat,
                                      aut)

In [11]:
# Criando uma nova coluna no "dataset", na qual corresponde à coluna "authors", porém
# substituindo os códigos por suas respectivas labels, disponíveis em "authors.csv"
dat["authors-parsed"] = authors_parsed

## `categories.csv` + `dataset.csv`: Substituindo index por label

In [12]:
# Lista com os nomes de categorias, dos quais subtituem os código de categorias
# da coluna "categories" do "dataset"
categories_parsed = parse_index_to_label("category_id", 
                                         "category_name", 
                                         "categories", 
                                         dat, 
                                         cat)

In [13]:
# Criando uma nova coluna no "dataset", na qual corresponde à coluna "categorie", porém
# substituindo os códigos por suas respectivas labels, disponíveis em "categories.csv"
dat["categories-parsed"] = categories_parsed

## `formats.csv` + `dataset.csv`: Substituindo index por label

In [14]:
# Mesma funcionalidade da função 'parse_index_to_label', porém adaptada
# para a coluna "format" do "dataset", da qual possui uma estrutura diferente.

# Inicializando a variável final com os itens substituídos
elements = []

# Iterando sobrea  coluna "format"
for i in dat["format"]:
    try:
        # Se conseguir converter o elemento para int, coleta e armazena o "name"
        k = fom[fom["format_id"] == int(i)]["format_name"].values[0]
    except:
        # Caso dê exceção, adiciona np.nan e pula para a próxima iteração
        elements.append(np.nan)
        continue

    # Adicona o elemento k, com o "name"
    elements.append(k)

# Lista com elementos substituídos
formats_parsed = elements

In [15]:
# Criando uma nova coluna no "dataset", na qual corresponde à coluna "format", porém
# substituindo os códigos por suas respectivas labels, disponíveis em "formats.csv"
dat["formats-parsed"] = formats_parsed

## Visualizando o DataFrame unificado: `dataset.csv`

In [16]:
# "dataset" após criação de novas colunas com mesclagem de dados
# Primeiras 5 linhas
dat.head()

authors  ...  formats-parsed
0  [1, 2]  ...       Paperback
1     [3]  ...       Paperback
2     [4]  ...       Paperback
3     [5]  ...       Paperback
4     [6]  ...       Paperback

[5 rows x 31 columns]

#Perguntas

## Qual a quantidade total de livros da base?

In [17]:
#Contando as ocorrências dos ids de livros da base dataset
dat.id.count()

20216

**Answer**: There are 20,216 books on the dataset.

##Qual a quantidade de livros que possui apenas 1 autor?


In [18]:
# Quantidade de ids dos livros no "dataset", excuíndo livros 
# com mais de um autor e vazios
len([i for i in dat.authors if "," not in i and i != "[]"])

14753

**Answer**: There are 14,753 books with one author.

##Quais os 5 autores com a maior quantidade de livros?


In [19]:
# Inicializando dicionário que conta a ocorrência de cada elemento
authors_counter = {}

for i in dat["authors-parsed"]:
    for k in i:
        authors_counter[k] = 0

# Incrementando o contador
for i in dat["authors-parsed"]:
    for k in i:
        authors_counter[k] += 1

In [20]:
# Ordenando o contador
sorted_authors_counter = {k: v for k, v in sorted(authors_counter.items(), 
                                                  key=lambda item: item[1])}

In [21]:
# Criando um DataFrame com o contador
df_books_per_authors = pd.DataFrame(sorted_authors_counter, 
                                    index=[0])

In [22]:
# Criando uma matriz transposta para fácil exibição e coletando os últimos
# 5 registros
np.transpose(df_books_per_authors).tail()

0
Kevin Lee             83
Matthew Preston       83
William Shakespeare  117
Rand McNally         212
Weezag               258

##Qual a quantidade de livros por categoria?


In [23]:
# Inicializando dicionário que conta a ocorrência de cada elemento
categories_counter = {}

for i in dat["categories-parsed"]:
    for k in i:
        categories_counter[k] = 0

# Incrementando o contador
for i in dat["categories-parsed"]:
    for k in i:
        categories_counter[k] += 1

In [24]:
# Ordenando o contador
sorted_categories_counter = {k: v for k, v in sorted(categories_counter.items(), 
                                                     key=lambda item: item[1])}

In [25]:
# Criando um DataFrame com o contador
df_books_per_categories = pd.DataFrame(sorted_categories_counter, 
                                    index=[0])

In [26]:
# Criando uma matriz transposta para fácil exibição
np.transpose(df_books_per_categories)

0
Zoos & Wildlife Parks                             1
Shipbuilding Technology, Engineering & Trades     1
Buses, Trams & Commercial                         1
Organic Farming                                   1
Engineering Skills & Trades                       1
...                                             ...
Theatre Studies                                1854
Usage & Grammar Guides                         1862
World Atlases / World Maps                     3792
Teaching Skills & Techniques                   4122
Television                                     4450

[1731 rows x 1 columns]

##Quais as 5 categorias com a maior quantidade de livros?


In [27]:
# Criando uma matriz transposta para fácil exibição e coletando os últimos
# 5 registros
np.transpose(df_books_per_categories).tail()

0
Theatre Studies               1854
Usage & Grammar Guides        1862
World Atlases / World Maps    3792
Teaching Skills & Techniques  4122
Television                    4450

##Qual o formato com a maior quantidade de livros?


In [28]:
# Inicializando dicionário que conta a ocorrência de cada elemento
formats_counter = {}

for i in dat["formats-parsed"]:
    formats_counter[i] = 0

# Incrementando o contador
for i in dat["formats-parsed"]:
    formats_counter[i] += 1

In [29]:
# Ordenando o contador
sorted_formats_counter = {k: v for k, v in sorted(formats_counter.items(), 
                                                  key=lambda item: item[1])}

In [30]:
# Criando um DataFrame com o contador
df_books_per_formats = pd.DataFrame(sorted_formats_counter, 
                                    index=[0])

In [31]:
# Criando uma matriz transposta para fácil exibição e coletando o último
# registro
np.transpose(df_books_per_formats).tail(1)

0
Paperback  13497

**Answer**: The format with the greatest number of books is "Paperblack".

##Considerando a coluna `bestsellers-rank`, quais os 10 livros mais bem posicionados?


In [32]:
acc = 1 # Contador

# Iterando sobre "bestsellers-rank", ordenando e coletando os 10 livros mais bem 
# posicionados
for i in dat["bestsellers-rank"].sort_values(ascending=False).index[:10]:
    print(f"{acc}. id: {dat.iloc[i].id} \t {dat.iloc[i].title}\n")
    
    acc += 1 # Incrementando

1. id: 9789089645227 	 After the Break : Television Theory Today

2. id: 9789053568439 	 On Latin Adverbs

3. id: 9789622096318 	 Feeling Asian Modernities - Transnational Consumption of Japanese TV Dramas

4. id: 9789042034709 	 Lexikon der alteren germanischen Lehnwoerter in den ostseefinnischen Sprachen, Band III : P-AE

5. id: 9789774249839 	 The Real (Arab) World: v. 2

6. id: 9789004194526 	 Multi-verb Constructions : A View from the Americas

7. id: 9789004177222 	 The Noun Phrase in Ancient Greek : A Functional Analysis of the Order and Articulation of NP Constituents in Herodotus

8. id: 9788865273135 	 Ghost Journal

9. id: 9789400738225 	 Handbook of Generative Approaches to Language Acquisition

10. id: 9789768200150 	 Ghost Stories of Arizona and New Mexico



##Considerando a coluna `rating-avg`, quais os 10 livros mais bem posicionados?


In [33]:
acc = 1 # Contador

# Iterando sobre "rating-avg", ordenando e coletando os 10 livros mais bem
# posicionados
for i in dat["rating-avg"].sort_values(ascending=False).index[:10]:
    print(f"{acc}. id: {dat.iloc[i].id} \t {dat.iloc[i].title}\n")
    
    acc += 1 # Incrementando

1. id: 9789020992922 	 Atlas Ferraris : de Eerste Atlas Van Belgie Le Premier Atlas de La Belgique

2. id: 9781789093247 	 Rick Baker: Metamorphosis

3. id: 9781257831944 	 Mammalogy Techniques Manual 2nd Edition

4. id: 9781476670775 	 Single Season Sitcoms of the 1990s : A Complete Guide

5. id: 9780880975988 	 New York City 5 Borough Atlas

6. id: 9780615234823 	 In Focus

7. id: 9781982235116 	 The Earth Keeper's Handbook : Assuming Leadership in a New World

8. id: 9781393589303 	 Collective Consciousness : How to Transcend Mass Consciousness and Become One With the Universe

9. id: 9780752438016 	 Scotch on Ice : Scotland's Ice Hockey Heroes

10. id: 9780813229737 	 The Business Francis Means : Understanding the Pope's Message on the Economy



##Quantos livros possuem `rating-avg` maior do que 3,5?


In [34]:
# Filtrando "rating-avg" maior que 3.5 e contando
dat[dat["rating-avg"] > 3.5].id.count()

8182

**Answer**: There are 8,182 books with "rating-avg" above 3.5.

##Quantos livros tem data de publicação (`publication-date`) maior do que `01-01-2020`?

In [35]:
# Filtrando "publication-date" após "2020-01-01" e contando
dat[dat["publication-date"] > "2020-01-01"].id.count()

1178

**Answer**: There are 1,178 books with "publication-date" above 2020/01/01.